proc_macros/src/lib.rs

```diff
@@ -165,3 +203,599 @@ pub fn rule_tests(input: TokenStream) -> TokenStream {
+            listeners: listeners.expect("Expected 'listeners'"),
+        })
+    }
+}
+
+#[proc_macro]
+pub fn rule(input: TokenStream) -> TokenStream {
+    let rule: Rule = parse_macro_input!(input);
+
+    let rule_struct_name = format_ident!("{}", rule.name_string().to_class_case());
+
+    let rule_state_fields = rule.state.as_ref().map_or_else(
+        || Default::default(),
+        |state| {
+            state
+                .scope_sections
+                .iter()
+                .filter(|scope_section| scope_section.scope == RuleStateScope::RuleStatic)
+                .flat_map(|scope_section| scope_section.fields.iter())
+                .collect::<Vec<_>>()
+        },
+    );
+
+    let rule_struct_definition = get_rule_struct_definition(&rule_struct_name, &rule_state_fields);
+
+    let rule_instance_struct_name = format_ident!("{}Instance", rule_struct_name);
+
+    let rule_instance_state_fields = rule.state.as_ref().map_or_else(
+        || Default::default(),
+        |state| {
+            state
+                .scope_sections
+                .iter()
+                .filter(|scope_section| scope_section.scope == RuleStateScope::PerRun)
+                .flat_map(|scope_section| scope_section.fields.iter())
+                .collect::<Vec<_>>()
+        },
+    );
+
+    let rule_rule_impl = get_rule_rule_impl(
+        &rule,
+        &rule_struct_name,
+        &rule_instance_struct_name,
+        &rule_instance_state_fields,
+    );
+
+    let rule_instance_struct_definition = get_rule_instance_struct_definition(
+        &rule_struct_name,
+        &rule_instance_struct_name,
+        &rule_instance_state_fields,
+    );
+
+    let rule_instance_per_file_struct_name = format_ident!("{}PerFile", rule_instance_struct_name);
+
+    let rule_instance_per_file_state_fields = rule.state.as_ref().map_or_else(
+        || Default::default(),
+        |state| {
+            state
+                .scope_sections
+                .iter()
+                .filter(|scope_section| scope_section.scope == RuleStateScope::PerFileRun)
+                .flat_map(|scope_section| scope_section.fields.iter())
+                .collect::<Vec<_>>()
+        },
+    );
+
+    let rule_instance_rule_instance_impl = get_rule_instance_rule_instance_impl(
+        &rule_instance_struct_name,
+        &rule_instance_per_file_struct_name,
+        &rule_instance_per_file_state_fields,
+    );
+
+    let rule_instance_per_file_struct_definition = get_rule_instance_per_file_struct_definition(
+        &rule_instance_struct_name,
+        &rule_instance_per_file_struct_name,
+        &rule_instance_per_file_state_fields,
+    );
+
+    let rule_instance_per_file_rule_instance_per_file_impl =
+        get_rule_instance_per_file_rule_instance_per_file_impl(
+            &rule,
+            &rule_instance_per_file_struct_name,
+        );
+
+    let instantiate_rule = get_rule_struct_creation(&rule_struct_name, &rule_state_fields);
```

Comment:
It seems like the desired usage pattern is that the macro immediately "returns" an "instance" of the newly defined rule ("instance"/"instantiated" are getting overloaded since that has a rule-specific meaning, here I just mean "creating an instance of the `struct` that implements `Rule`")

So here we just assume that we're in an "expression context" (?) and "return" an `Arc<TheGeneratedRuleStruct>`